kubernetes/kubernetes.go

```diff
@@ -0,0 +1,223 @@
+	if err != nil {
+		return nil, err
+	}
+
+	endpoints, err := getEndpoints(clients.CoreV1().Endpoints(""))
+	if err != nil {
+		return nil, err
+	}
+
+	return mergeBackends(ingresses, services, endpoints), nil
+}
+
+func mergeBackends(ingresses map[HostMatch]Ingress, services map[PortMatch]Service, endpoints map[Object][]Endpoint) map[string][]util.Backend {
+
+	backends := make(map[string][]util.Backend)
+	for n, i := range ingresses {
+		backends[n.HostName] = toBackendList(i.Scheme, services[PortMatch{Object: n.Object, Port: i.Port}], endpoints[n.Object])
+	}
+
+	return backends
+}
+
+func toBackendList(scheme string, service Service, endpoints []Endpoint) []util.Backend {
+	backends := make([]util.Backend, 0)
+	for _, e := range endpoints {
+		if e.Port == service.TargetPort {
+			backends = append(backends, util.Backend{
+				Url: &url.URL{
+					Scheme: scheme,
+					Host:   e.Address + ":" + strconv.FormatInt(int64(e.Port), 10),
+				},
+			})
+		}
+	}
+
+	return backends
+}
+
+func getServices(client v12.ServiceInterface) (map[PortMatch]Service, error) {
+	services, err := client.List(v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(map[PortMatch]Service)
+	for _, s := range services.Items {
+		for n, ss := range mapService(s) {
+			out[n] = ss
+		}
+	}
+
+	return out, nil
+}
+
+func mapService(service v13.Service) map[PortMatch]Service {
+	out := make(map[PortMatch]Service)
+	for _, s := range service.Spec.Ports {
+		var sourcePort, targetPort uint16
+		sourcePort = i32toPort(s.Port)
+		if i32toPort(s.TargetPort.IntVal) > 0 {
+			targetPort = i32toPort(s.TargetPort.IntVal)
+		} else {
+			targetPort = sourcePort
+		}
+
+		if sourcePort > 0 && targetPort > 0 {
+			out[PortMatch{
+				Object: Object{
+					Name:      service.Name,
+					Namespace: service.Namespace,
+				},
+				Port: sourcePort,
+			}] = Service{
+				Port:       uint16(s.Port),
+				TargetPort: targetPort,
+			}
+		}
+	}
+
+	return out
+}
+
+func getEndpoints(client v12.EndpointsInterface) (map[Object][]Endpoint, error) {
+	endpoints, err := client.List(v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(map[Object][]Endpoint)
+	for _, e := range endpoints.Items {
+		out[Object{Name: e.Name, Namespace: e.Namespace}] = mapEndpoint(e)
+	}
+
+	return out, nil
+}
+
+func getIngresses(client v1beta1.IngressInterface) (map[HostMatch]Ingress, error) {
+	ingresses, err := client.List(v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(map[HostMatch]Ingress)
+	for _, i := range ingresses.Items {
+		for host, backend := range mapIngress(i) {
+			out[HostMatch{
+				Object:   Object{Name: backend.Name, Namespace: i.Namespace},
+				HostName: host,
+			}] = backend
+		}
+	}
+
+	return out, nil
+}
+
+func mapEndpoint(in v13.Endpoints) []Endpoint {
+	out := make([]Endpoint, 0)
+	for _, s := range in.Subsets {
+		for _, p := range s.Ports {
+			if p.Protocol == v13.ProtocolTCP && i32toPort(p.Port) > 0 {
```

Comment:
So we assume it's HTTP if the protocol is TCP?

Reply:
No? We just know that HTTP is _always_ TCP. And since this proxy only handles HTTP it's safe to always drop everything that's not TCP.